contracts/defi/provider/CompoundV2Provider.sol

```diff
@@ -0,0 +1,405 @@
+        uint exchangeRateMantissa = CToken(cToken).exchangeRateStored();
+        _tokenValue = amount.mul(exchangeRateMantissa).div(10 ** 18);
+        _periodEnd = 0;
+    }
+
+    /* ********************************** Implementation of Loan ************************************* */
+
+    /**
+     * @dev Opens a collateralized loan.
+     * @param _wallet The target wallet.
+     * @param _collateral The token used as a collateral.
+     * @param _collateralAmount The amount of collateral token provided.
+     * @param _debtToken The token borrowed.
+     * @param _debtAmount The amount of tokens borrowed.
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     * @return (optional) An ID for the loan when the provider enables users to create multiple distinct loans.
+     */
+    function openLoan(
+        BaseWallet _wallet, 
+        address _collateral, 
+        uint256 _collateralAmount, 
+        address _debtToken, 
+        uint256 _debtAmount, 
+        address[] calldata _oracles
+    ) 
+        external 
+        returns (bytes32 _loanId) 
+    {
+        require(_oracles.length == 2, "CompoundV2: invalid oracles length");
+        address[] memory markets = new address[](2);
+        markets[0] = CompoundRegistry(_oracles[1]).getCToken(_collateral);
+        markets[1] = CompoundRegistry(_oracles[1]).getCToken(_debtToken);
+        address comptroller = _oracles[0];
+        _wallet.invoke(comptroller, 0, abi.encodeWithSignature("enterMarkets(address[])", markets));
+        mint(_wallet, markets[0], _collateral, _collateralAmount);
+        borrow(_wallet, markets[1], _debtAmount);
+    }
+
+    /**
+     * @dev Closes a collateralized loan by repaying all debts (plus interest). Note that it does not redeem the collateral.
+     * @param _wallet The target wallet.
+     * @param _loanId The ID of the loan if any, 0 otherwise.
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     */
+    function closeLoan(
+        BaseWallet _wallet, 
+        bytes32 _loanId, 
+        address[] calldata _oracles
+    ) 
+        external 
+    {
+        require(_oracles.length == 2, "CompoundV2: invalid oracles length");
+        address comptroller = _oracles[0];
+        address[] memory markets = Comptroller(comptroller).getAssetsIn(address(_wallet));
+        for(uint i = 0; i < markets.length; i++) {
+            address cToken = markets[i];
+            uint debt = CToken(cToken).borrowBalanceCurrent(address(_wallet));
+            if(debt > 0) { 
+                repayBorrow(_wallet, cToken, debt);
+                uint collateral = CToken(cToken).balanceOf(address(_wallet));
+                if(collateral == 0) { 
+                    _wallet.invoke(comptroller, 0, abi.encodeWithSignature("exitMarket(address)", address(cToken)));
```

Comment:
I understand that it does not hurt to come back to the initial state of not being in the market but is there any other benefit to exiting the market when there is no collateral left in it?

Reply:
Yes, because the Comptroller checks the status of each entered market when the loan is modified (collateral/debt increased/decreased) which costs gas. So exiting an empty market reduces the gas of every operation.